### PR TITLE
[IA-4935] Generate azure prod vm password randomly

### PR DESCRIPTION
**Pull request by LizBaldo**
__Jira ticket__: https://broadworkbench.atlassian.net/browse/[ticket_number]

### What

- Replaces the Azure VM password with a randomly generated one in prod

### Why

- Sharing a single password was a high security risk that needs to be mitigated

### What to test

I will test on my BEE, switching the logic to check for the random password there, but when it comes to testing if it works in prod... Well I am not quite sure how I can mock that 😅

### CODE REVIEW

File: http/src/main/resources/leo.conf [hunk 1]
```diff
@@ -4,6 +4,7 @@ application {
   leoServiceAccountJsonFile = ${?LEO_SA_JSON_PATH}
   leoServiceAccountEmail = ${?LEO_SVC_ACCT_EMAIL}
   leoUrlBase = ${?LEO_URL_BASE}
+  environment = ${?APP_ENVIRONMENT}
```
Comment (LizBaldo) [inline http/src/main/resources/leo.conf:7]: It was already in the leo helm values 🎉

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala [hunk 2]
```diff
@@ -164,6 +166,35 @@ class AzurePubsubHandlerInterp[F[_]: Parallel](
       )
     } yield ()
 
+  private def generateAzureVMSecurePassword(): String = {
```
Comment (LizBaldo) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: I kinda wanted to add a unit test for this, but not quite sure if this is the best location for this function in the first place. There does not seem to be a kind of `utils` module for leo that i could stick it into

Comment (jdcanas) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: I heard an interesting philosophy recently that `util` files are generally a smell and are an easy way to end up with dependency spaghetti (*points to UI dependency graph which has tons of utils*). Rather, it is better for util methods to live in the file they are consumed, especially if there is only one such file. I think its fine here, and fine to test directly (though you'll need `private[util]`

Comment (jdcanas) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: Also I just realized this is totally in a `util` folder 🤣

**Review by jdcanas — APPROVED** [review]
Looks good once unit tested (perhaps assert the azure password requirements and length?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala (hunk 2 above)
Comment (cpate4) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: (in the interest of adding to this conversation) - I agree with @jdcanas re: smell (the single file ends up being a dumping ground), and also, I've also read that an alternative is to having a single file is to break out the file into parts and have one "util" aptly named file that reflects the method it represents.

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandlerAlgebra.scala [hunk 3]
```diff
@@ -144,3 +147,33 @@ final case class AzurePubsubHandlerConfig(samUrl: Uri,
 ) {
   def welderImage: String = s"$welderAcrUri:$welderImageHash"
 }
+object AzurePubsubHandler {
+  private[util] def generateAzureVMSecurePassword(): String = {
+    // Azure is enforcing the following constraints for password generation
+    // Passwords must not include reserved words or unsupported characters.
+    // Password must have 3 of the following: 1 lower case character, 1 upper case character, 1 number, and 1 special character that is not '\'or '-'.
+    // The value must be between 12 and 123 characters long.
```
Comment (cpate4) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandlerAlgebra.scala:155]: just confirming that:
- the above is the requirements, and
- we have intentionally settled on 16char for our random password length (vs having random 12-to-123 characters long)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala [hunk 4]
```diff
@@ -228,6 +228,12 @@ class AzurePubsubHandlerInterp[F[_]: Parallel](
       .minorVersionAutoUpgrade(config.runtimeDefaults.customScriptExtension.minorVersionAutoUpgrade)
       .protectedSettings(protectedSettings.asJava)
 
+    // Generate random password for Azure VM in production, for the other lower level envs we can used the shared password
+    val vmPassword = applicationConfig.environment match {
+      case "prod" => AzurePubsubHandler.generateAzureVMSecurePassword()
+      case _      => config.runtimeDefaults.vmCredential.password
+    }
```
Comment (cpate4) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: my two cents ... I would put this logic in a method, and write a test to show `prod` behaves as we expect, and non-prod results in the expected password, and one showing the method is called where we expect it to be (here). 

those would be the only additional tests I would write, as opposed to figuring out integration tests to etc. etc. etc.

**Review by mlilynolting — APPROVED** [review]
Looks good to me, with the common-sense unit test :)

**Review by cpate4 — APPROVED** [review]
couple knowledge comments and just my two-cents re: testing. integration tests are good, but I am not entirely sure we need one here. happy to support either approach (#LLL - Long Live Leonardo - 🤓)

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala (hunk 4 above)
Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: Good point, let me give this a try, thanks!

Comment (LizBaldo) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/AzurePubsubHandler.scala]: Test added, and I was able to use the randomly generated password to create an Azure VM from the portal a couple of times, so we should be good to go

**Comment by LizBaldo** [comment]
Going to force merge, this is an Azure change that has nothing to do with the GKE app creation failure that is a known flake

**Pull request merged**